emission/analysis/modelling/trip_model/gradient_boosted_decision_tree.py

```diff
@@ -0,0 +1,122 @@
+        for k in expected_keys:
+            if config.get(k) is None:
+                msg = f"gbdt trip model config missing expected key {k}"
+                raise KeyError(msg)
+        self.is_incremental = config['incremental_evaluation']
+        # Use the sklearn implementation of a GBDT
+        self.gbdt = GradientBoostingClassifier(n_estimators=50)
+        # Which features to use in the fit/prediction
+        self.feature_list = config['feature_list']
+        self.dependent_var = config['dependent_var']
+
+    def fit(self, trips: List[ecwc.Confirmedtrip]):
+        """train the model by passing data, where each row in the data
+        corresponds to a label at the matching index of the label input
+
+        :param trips: 2D array of features to train from
+        """
+        logging.debug(f'fit called with {len(trips)} trips')
+        unlabeled = list(filter(lambda t: len(t['data']['user_input']) == 0, trips))
+        if len(unlabeled) > 0:
+            msg = f'model.fit cannot be called with unlabeled trips, found {len(unlabeled)}'
+            raise Exception(msg)
+        X_train, y_train = self.extract_features(trips)
+        self.gbdt.fit(X_train, y_train)
+        logging.info(f"gradient boosted decision tree model fit to {len(X_train)} rows of trip data")
+
+    def predict(self, trip: ecwc.Confirmedtrip) -> Tuple[List[Dict], int]:
+        logging.debug(f"running gradient boosted mode prediction")
+        X_train, y_train = self.extract_features(trip)
+        y_pred = self.gbdt.predict(X_train)
+        if y_pred is None:
+            logging.debug(f"unable to predict bin for trip {trip}")
+            return []
+        else:
+            logging.debug(f"made predictions {y_pred}")
+            return y_pred
+
+    def to_dict(self) -> Dict:
+        return self.gbdt
+
+    def from_dict(self, model: Dict):
+        self.gbdt = model
+
+    def extract_features(self, trips: ecwc.Confirmedtrip) -> List[float]:
+        # TODO: need to enable generic paths other than just user input for features
+        X = pd.DataFrame(
+            [[trip['data']['user_input'][x] for x in self.feature_list] for trip in trips],
+            columns=self.feature_list
+        )
+        y = pd.DataFrame(
+            [trip['data']['user_input'][self.dependent_var] for trip in trips],
+            columns=[self.dependent_var]
+        )
+        # Clean up and recode the feature columns for training/prediction
+        X_processed, y_processed = self._process_data(X, y)
+        return X_processed, y_processed
+
+    def _process_data(self, X, y):
+        """
+        helper function to transform binned features and labels.
+        """
+        # Any non-numeric dtype must be one-hot encoded (if unordered) or numerically coded (if ordered)
+        dummies = []
+        for col in X:
+            if X[col].dtype=='object':
+                dummies.append(pd.get_dummies(X[col], prefix=col))
+        X = pd.concat(dummies, axis=1)
```

Comment:
Interesting, I did not know about `pandas.get_dummies` before. 
What are the columns that are affected by this?

Reply:
My thought is that this should work on strings, dates, mixed columns, anything unordinary but especially columns that are categorical/strings, which will not work as features for sklearn. I automatically convert them to dummy variables, assuming that if they were part of the specification, they were put there intentionally. I could see this going wrong if a column was specified that had many different responses such as a user input which would lead to a very large, sparse set of dummy variables. However the model would still train and not break. It may also lead to the same effect if there were a numeric column that had one string in it or something along those lines.

Reply:
Good job adding a related comment in the code.